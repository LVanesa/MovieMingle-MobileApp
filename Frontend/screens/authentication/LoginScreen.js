import React, { useState, useContext } from "react";
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  ImageBackground,
  KeyboardAvoidingView,
  ScrollView,
  Platform,
  Alert,
} from "react-native";
import { useNavigation } from "@react-navigation/native";
import { loginUser } from "../../api/authApi";
import { AuthContext } from "../../store/auth-context";
import InputField from "../../components/InputField";
//import { useGoogleAuth } from "./GoogleAuth";

const LoginScreen = () => {
  const navigation = useNavigation();
  const authCtx = useContext(AuthContext);

  const [form, setForm] = useState({ email: "", password: "" });
  const [errors, setErrors] = useState({});

  //const { promptAsync } = useGoogleAuth();

  const handleChange = (key, value) => {
    setForm((prev) => ({ ...prev, [key]: value }));
    setErrors((prev) => ({ ...prev, [key]: null, general: null }));
  };

  const validateForm = () => {
    const newErrors = {};
    if (!form.email.trim()) newErrors.email = "Email is required";
    if (!form.password.trim()) newErrors.password = "Password is required";
    return newErrors;
  };

  const handleLogin = async () => {
    setErrors({}); // Clear all errors before validation

    const validationErrors = validateForm();
    if (Object.keys(validationErrors).length > 0) {
      setErrors(validationErrors);
      return;
    }

    try {
      const response = await loginUser(form);
      const token = response.data.token;
      authCtx.authenticate(token);
      Alert.alert("Success", "You're now logged in!");
    } catch (error) {
      const errorMsg =
        error?.response?.data?.error ||
        "Login failed. Please check your credentials.";
      setErrors({ general: errorMsg }); // Clear field errors, only show general error
    }
  };

  return (
    <ImageBackground
      source={require("../../assets/images/main-background.jpg")}
      style={styles.background}
    >
      <KeyboardAvoidingView
        behavior={Platform.OS === "ios" ? "padding" : undefined}
        style={{ flex: 1 }}
      >
        <ScrollView contentContainerStyle={styles.container}>
          <View style={styles.card}>
            <Text style={styles.title}>Login</Text>

            {errors.general && (
              <Text testID="errorMessage" style={styles.generalError}>
                {errors.general}
              </Text>
            )}

            <InputField
              testID="emailInput"
              accessibilityLabel="Email input"
              label="Email"
              value={form.email}
              onChange={(text) => handleChange("email", text)}
              placeholder="Enter your email"
              keyboardType="email-address"
              error={errors.email}
            />

            <InputField
              testID="passwordInput"
              accessibilityLabel="Password input"
              label="Password"
              value={form.password}
              onChange={(text) => handleChange("password", text)}
              placeholder="Enter your password"
              secureTextEntry
              error={errors.password}
            />

            <TouchableOpacity style={styles.forgotPassword}>
              <Text style={styles.linkUnderline}>Forgot Password?</Text>
            </TouchableOpacity>

            <TouchableOpacity
              style={styles.button}
              onPress={handleLogin}
              testID="loginButton"
              accessibilityLabel="Login button"
            >
              <Text style={styles.buttonText}>Log in</Text>
            </TouchableOpacity>

            {/* <View style={styles.separator}>
              <View style={styles.line} />
              <Text style={styles.separatorText}>OR</Text>
              <View style={styles.line} />
            </View>

            <TouchableOpacity
              style={styles.googleButton}
              onPress={() => promptAsync()}
            >
              <Text style={styles.googleText}>Login with Google</Text>
            </TouchableOpacity>
 */}

            <TouchableOpacity
              onPress={() => navigation.navigate("Register")}
              style={{ marginTop: 15 }}
            >
              <Text style={styles.link}>
                Donâ€™t have an account?{" "}
                <Text style={styles.linkUnderline}>Register!</Text>
              </Text>
            </TouchableOpacity>
          </View>
        </ScrollView>
      </KeyboardAvoidingView>
    </ImageBackground>
  );
};

const styles = StyleSheet.create({
  background: {
    flex: 1,
    resizeMode: "cover",
  },
  container: {
    flexGrow: 1,
    justifyContent: "center",
    alignItems: "center",
    padding: 20,
  },
  card: {
    width: "100%",
    maxWidth: 400,
    backgroundColor: "rgba(0,0,0,0.6)",
    borderRadius: 20,
    padding: 20,
  },
  title: {
    fontSize: 28,
    color: "white",
    textAlign: "center",
    marginBottom: 20,
  },
  forgotPassword: {
    alignSelf: "flex-end",
    marginBottom: 10,
  },
  link: {
    color: "#fff",
    fontSize: 14,
    textAlign: "center",
  },
  linkUnderline: {
    textDecorationLine: "underline",
    fontWeight: "bold",
    color: "#fff",
  },
  button: {
    backgroundColor: "#fff",
    borderRadius: 40,
    paddingVertical: 12,
    alignItems: "center",
    marginTop: 10,
  },
  buttonText: {
    color: "#000",
    fontSize: 16,
    fontWeight: "bold",
  },
  separator: {
    flexDirection: "row",
    alignItems: "center",
    marginVertical: 15,
  },
  separatorText: {
    color: "#fff",
    marginHorizontal: 10,
  },
  line: {
    flex: 1,
    height: 1,
    backgroundColor: "#ccc",
  },
  googleButton: {
    backgroundColor: "#4285F4",
    borderRadius: 40,
    paddingVertical: 12,
    alignItems: "center",
  },
  googleText: {
    color: "#fff",
    fontSize: 16,
    fontWeight: "600",
  },
  generalError: {
    color: "#FF6B6B",
    textAlign: "center",
    marginBottom: 12,
  },
});

export default LoginScreen;
