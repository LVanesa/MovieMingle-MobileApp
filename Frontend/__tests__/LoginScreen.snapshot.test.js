import { render } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";
jest.mock("@react-native-async-storage/async-storage", () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}));
import LoginScreen from "../screens/authentication/LoginScreen";
import React from "react";

test("renders login screen correctly", () => {
  const tree = render(
    <NavigationContainer>
      <LoginScreen />
    </NavigationContainer>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
