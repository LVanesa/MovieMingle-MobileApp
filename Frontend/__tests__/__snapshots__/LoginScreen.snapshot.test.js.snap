// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders login screen correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    {
      "flex": 1,
      "resizeMode": "cover",
    }
  }
>
  <Image
    source={
      {
        "testUri": "../../../assets/images/main-background.jpg",
      }
    }
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "alignItems": "center",
          "flexGrow": 1,
          "justifyContent": "center",
          "padding": 20,
        }
      }
    >
      <View>
        <View
          style={
            {
              "backgroundColor": "rgba(0,0,0,0.6)",
              "borderRadius": 20,
              "maxWidth": 400,
              "padding": 20,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 28,
                "marginBottom": 20,
                "textAlign": "center",
              }
            }
          >
            Login
          </Text>
          <View
            style={
              {
                "marginBottom": 12,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "marginBottom": 4,
                }
              }
            >
              Email
            </Text>
            <TextInput
              accessibilityLabel="Email input"
              autoCapitalize="none"
              keyboardType="email-address"
              onChangeText={[Function]}
              placeholder="Enter your email"
              secureTextEntry={false}
              style={
                [
                  {
                    "backgroundColor": "#fff",
                    "borderRadius": 8,
                    "fontSize": 15,
                    "paddingHorizontal": 12,
                    "paddingVertical": 10,
                  },
                  undefined,
                ]
              }
              testID="emailInput"
              value=""
            />
          </View>
          <View
            style={
              {
                "marginBottom": 12,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "marginBottom": 4,
                }
              }
            >
              Password
            </Text>
            <TextInput
              accessibilityLabel="Password input"
              autoCapitalize="none"
              keyboardType="default"
              onChangeText={[Function]}
              placeholder="Enter your password"
              secureTextEntry={true}
              style={
                [
                  {
                    "backgroundColor": "#fff",
                    "borderRadius": 8,
                    "fontSize": 15,
                    "paddingHorizontal": 12,
                    "paddingVertical": 10,
                  },
                  undefined,
                ]
              }
              testID="passwordInput"
              value=""
            />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignSelf": "flex-end",
                "marginBottom": 10,
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "textDecorationLine": "underline",
                }
              }
            >
              Forgot Password?
            </Text>
          </View>
          <View
            accessibilityLabel="Login button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 40,
                "marginTop": 10,
                "opacity": 1,
                "paddingVertical": 12,
              }
            }
            testID="loginButton"
          >
            <Text
              style={
                {
                  "color": "#000",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              Log in
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "marginTop": 15,
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 14,
                  "textAlign": "center",
                }
              }
            >
              Donâ€™t have an account?
               
              <Text
                style={
                  {
                    "color": "#fff",
                    "fontWeight": "bold",
                    "textDecorationLine": "underline",
                  }
                }
              >
                Register!
              </Text>
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
